test/liquid_voting/delegations_test.exs

```diff
@@ -61,24 +64,47 @@ defmodule LiquidVoting.DelegationsTest do
       assert {:ok, %Delegation{} = delegation} = Delegations.create_delegation(args)
     end
 
-    test "create_delegation/1 with proposal url sets global boolean to false", context do
-      proposal_url = "https://www.someorg/proposalX"
+    test "create_delegation/1 with duplicate data returns error changeset", context do
+      Delegations.create_delegation(context[:valid_attrs])
+      assert {:error, %Ecto.Changeset{}} = Delegations.create_delegation(context[:valid_attrs])
+    end
 
-      args = Map.merge(context[:valid_attrs], %{proposal_url: proposal_url})
-      {:ok, %Delegation{} = delegation} = Delegations.create_delegation(args)
+    test "upsert_delegation/1 with duplicate delegator and proposal_url updates the respective delegation",
+         context do
+      args = Map.merge(context[:valid_attrs], %{proposal_url: @proposal_url})
+      {:ok, %Delegation{} = original_delegation} = Delegations.create_delegation(args)
```

Comment:
Better to use a [factory](https://github.com/thoughtbot/ex_machina) here. This separates setup better from code under test, and code will be more concise. Also here specifically we avoid another write to the db - `context[:valid_attrs]` has values from another factory inserted on `setup()` above.

Factories can have custom fields too, as in `insert(:delegation, proposal_url: '...')`

Reply:
Is this done?

Reply:
I think so. I created and used 
```
valid_proposal_specific_attrs: %{
          delegator_id: delegator.id,
          delegate_id: delegate.id,
          organization_id: organization_id,
          proposal_url: proposal_url
        },
        update_proposal_specific_attrs: %{
          delegator_id: delegator.id,
          delegate_id: another_delegate.id,
          organization_id: organization_id,
          proposal_url: proposal_url
        }
```
Which I assumed == using factories.

Reply:
It's certainly a cleaner approach.

Reply:
Ignore what I said earlier about the same delegation being created twice, I had mis-read the setup code.

What I had in mind was this:

```elixir
# factory-created original delegation and new delegate, explicitly not code under test
delegation = insert(:delegation)
new_delegate = insert(:participant)

args = {% 
  delegator_id: delegation.delegator_id,
  delegate_id: new_delegate.id,
  organization_id: valid_attrs[:organization_id]
}
assert {:ok, %Delegation{} = modified_delegation} = Delegations.upsert_delegation(args)
assert modified_delegation.id == delegation.id
assert modified_delegation.organization_id == delegation.organization_id
assert modified_delegation.delegate_id != delegation.delegate_id
assert modified_delegation.delegator_id == delegation.delegator_id

# no need to list delegations and count them, since we've already asserted the same record id
```

Your latest update did make things clearer, though. Just wondering whether setup is getting bloated. Perhaps we can leave it as is, and refactor the test file later.

Your call. Only hard request here is to replace listing delegations to assert count, and just assert on the same record id.

Reply:
I can remove the count, but I didn't because  you said "I don't think we need to test the delegation count here, since we're already testing it on the LiquidVoting.Delegations layer." in a comment about the absinthe tests.

I don't think we need to test the delegation count here, since we're already testing it on the LiquidVoting.Delegations layer.

Reply:
Sure, but asserting the id ensure the count is 1 as well. It's the same record and so the count didn't change

Reply:
Got it. I'll remove it after my pairing.

Reply:
Another thing to keep in mind is, it's a good practice for unit tests to exercise as little code external to actual code under test as possible.

Asserting the id is more reliable since it doesn't depend on another function. We're just asserting over the persisted record, which is an outcome of the upsert itself. If `Delegations.list_delegations` gets a bug, it won't affect it.

Same reasoning applies to using factories instead of using `Delegations.create_delegation`, because the factory just populates the db. Currently if `Delegations.create_delegation` gets a bug down the line, this test for upserts will still run, giving us a false negative (or positive, depending on how one sees it). A false breakage detection: upsert actually works, but its test breaks because we're coupling it with an external function to the upsert.

So I guess I am requesting for `Delegations.create_delegation` to be replaced with a factory, sorry man =D

Reply:
(no rush: async, etc)

Reply:
+1 on leveraging `ex_machina` in this case as long as we have integration tests covering the _units_ together.

Reply:
No problem. I think I initially misunderstood what you wanted. I _think_ I get it now.

Reply:
How do add a `proposal_url` field + value to `insert(:delegation)` factory item?

Reply:
> No problem. I think I initially misunderstood what you wanted. I think I get it now.

and I probably was a little convoluted in my explaining it 😄 this discussion on tests on Friday will be good. And in general we should pair more to avoid these long running reviews, which can be taxing

Reply:
done: removed delegations counts + use factories to replace calls to create_delegations/1